File: patterns/_tests/BehavioralPatterns/Interpreter/ruby/unit.test.rb
# ============================================================
#  Unit Tests — Interpreter: Arithmetic Expression Evaluator (Ruby)
#
#  Run: ruby unit.test.rb
# ============================================================

require 'minitest/autorun'
require_relative '../../../../BehavioralPatterns/Interpreter/ruby/ruby'


# ── Context ───────────────────────────────────────────────────
class TestContext < Minitest::Test
  def test_stores_and_retrieves_variable
    ctx = Context.new.set("x", 42)
    assert_equal 42, ctx.get("x")
  end

  def test_raises_on_missing_variable
    assert_raises(KeyError) { Context.new.get("missing") }
  end

  def test_overwrites_existing_variable
    ctx = Context.new.set("x", 1).set("x", 99)
    assert_equal 99, ctx.get("x")
  end

  def test_set_returns_self_for_chaining
    ctx = Context.new
    assert_same ctx, ctx.set("a", 1)
  end
end


# ── NumberExpression ──────────────────────────────────────────
class TestNumberExpression < Minitest::Test
  def test_returns_literal_value
    assert_equal 7, NumberExpression.new(7).interpret(Context.new)
  end

  def test_ignores_context
    ctx = Context.new.set("x", 999)
    assert_equal 3, NumberExpression.new(3).interpret(ctx)
  end

  def test_display_renders_integer
    assert_equal "5", NumberExpression.new(5).display
  end
end


# ── VariableExpression ────────────────────────────────────────
class TestVariableExpression < Minitest::Test
  def test_resolves_from_context
    ctx = Context.new.set("a", 10)
    assert_equal 10, VariableExpression.new("a").interpret(ctx)
  end

  def test_raises_when_not_in_context
    assert_raises(KeyError) { VariableExpression.new("z").interpret(Context.new) }
  end

  def test_display_returns_name
    assert_equal "a", VariableExpression.new("a").display
  end
end


# ── AddExpression ─────────────────────────────────────────────
class TestAddExpression < Minitest::Test
  def test_adds_two_numbers
    expr = AddExpression.new(NumberExpression.new(3), NumberExpression.new(4))
    assert_equal 7, expr.interpret(Context.new)
  end

  def test_adds_two_variables
    ctx  = Context.new.set("a", 10).set("b", 5)
    expr = AddExpression.new(VariableExpression.new("a"), VariableExpression.new("b"))
    assert_equal 15, expr.interpret(ctx)
  end

  def test_display_wraps_in_parens
    expr = AddExpression.new(NumberExpression.new(1), NumberExpression.new(2))
    assert_includes expr.display, "+"
    assert expr.display.start_with?("(")
  end
end


# ── SubtractExpression ────────────────────────────────────────
class TestSubtractExpression < Minitest::Test
  def test_subtracts_right_from_left
    expr = SubtractExpression.new(NumberExpression.new(10), NumberExpression.new(3))
    assert_equal 7, expr.interpret(Context.new)
  end

  def test_result_can_be_negative
    expr = SubtractExpression.new(NumberExpression.new(2), NumberExpression.new(5))
    assert_equal(-3, expr.interpret(Context.new))
  end
end


# ── MultiplyExpression ────────────────────────────────────────
class TestMultiplyExpression < Minitest::Test
  def test_multiplies_two_numbers
    expr = MultiplyExpression.new(NumberExpression.new(4), NumberExpression.new(5))
    assert_equal 20, expr.interpret(Context.new)
  end

  def test_multiplies_variable_by_number
    ctx  = Context.new.set("a", 6)
    expr = MultiplyExpression.new(VariableExpression.new("a"), NumberExpression.new(7))
    assert_equal 42, expr.interpret(ctx)
  end
end


# ── DivideExpression ──────────────────────────────────────────
class TestDivideExpression < Minitest::Test
  def test_divides_left_by_right
    expr = DivideExpression.new(NumberExpression.new(10), NumberExpression.new(2))
    assert_equal 5.0, expr.interpret(Context.new)
  end

  def test_raises_on_division_by_zero
    expr = DivideExpression.new(NumberExpression.new(5), NumberExpression.new(0))
    assert_raises(ZeroDivisionError) { expr.interpret(Context.new) }
  end

  def test_produces_fractional_result
    expr = DivideExpression.new(NumberExpression.new(10), NumberExpression.new(4))
    assert_in_delta 2.5, expr.interpret(Context.new), 0.001
  end
end


# ── Composite trees ───────────────────────────────────────────
class TestCompositeTrees < Minitest::Test
  def test_diff_of_squares
    a    = VariableExpression.new("a")
    b    = VariableExpression.new("b")
    ctx  = Context.new.set("a", 10).set("b", 5)
    expr = MultiplyExpression.new(AddExpression.new(a, b), SubtractExpression.new(a, b))
    assert_equal 75, expr.interpret(ctx)
  end

  def test_same_tree_different_contexts
    a    = VariableExpression.new("a")
    b    = VariableExpression.new("b")
    expr = MultiplyExpression.new(AddExpression.new(a, b), SubtractExpression.new(a, b))
    assert_equal 75, expr.interpret(Context.new.set("a", 10).set("b", 5))
    assert_equal 32, expr.interpret(Context.new.set("a",  6).set("b", 2))
  end

  def test_literal_three_level_tree
    expr = MultiplyExpression.new(
      AddExpression.new(NumberExpression.new(3), NumberExpression.new(4)),
      NumberExpression.new(2)
    )
    assert_equal 14, expr.interpret(Context.new)
  end

  def test_nested_display_wraps_each_level
    expr = AddExpression.new(
      MultiplyExpression.new(NumberExpression.new(2), NumberExpression.new(3)),
      NumberExpression.new(1)
    )
    assert_includes expr.display, "(2 * 3)"
  end
end

File: patterns/_tests/BehavioralPatterns/Interpreter/python/unit.test.py
# ============================================================
#  Unit Tests - Interpreter: Arithmetic Expression Evaluator
#
#  Run: pytest unit.test.py -v
# ============================================================

import sys, os, math
sys.path.insert(0, os.path.join(os.path.dirname(__file__),
                                '../../../../BehavioralPatterns/Interpreter/python'))
from python import (
    Context,
    NumberExpression, VariableExpression,
    AddExpression, SubtractExpression, MultiplyExpression, DivideExpression,
    FloorDivideExpression, ModuloExpression, PowerExpression,
    NegateExpression, AbsExpression, SqrtExpression, InverseExpression,
    FactorialExpression, PowerOfTenExpression, LogExpression,
)
import pytest

CTX = Context()


# ── Context ───────────────────────────────────────────────────
class TestContext:
    def test_stores_and_retrieves(self): assert Context().set("x", 42).get("x") == 42
    def test_raises_on_missing(self):
        with pytest.raises(KeyError): Context().get("z")
    def test_overwrites(self): assert Context().set("x", 1).set("x", 99).get("x") == 99
    def test_set_returns_self(self):
        ctx = Context(); assert ctx.set("a", 1) is ctx


# ── Terminal expressions ──────────────────────────────────────
class TestNumberExpression:
    def test_returns_value(self): assert NumberExpression(7).interpret(CTX) == 7
    def test_ignores_context(self): assert NumberExpression(3).interpret(Context().set("x", 99)) == 3
    def test_str_integer(self): assert str(NumberExpression(5)) == "5"

class TestVariableExpression:
    def test_resolves(self): assert VariableExpression("a").interpret(Context().set("a", 10)) == 10
    def test_raises_missing(self):
        with pytest.raises(KeyError): VariableExpression("z").interpret(CTX)
    def test_str_name(self): assert str(VariableExpression("a")) == "a"


# ── Binary operators ──────────────────────────────────────────
class TestAddExpression:
    def test_adds(self): assert AddExpression(NumberExpression(3), NumberExpression(4)).interpret(CTX) == 7

class TestSubtractExpression:
    def test_subtracts(self): assert SubtractExpression(NumberExpression(10), NumberExpression(3)).interpret(CTX) == 7
    def test_negative(self):  assert SubtractExpression(NumberExpression(2), NumberExpression(5)).interpret(CTX) == -3

class TestMultiplyExpression:
    def test_multiplies(self): assert MultiplyExpression(NumberExpression(4), NumberExpression(5)).interpret(CTX) == 20

class TestDivideExpression:
    def test_divides(self):   assert DivideExpression(NumberExpression(10), NumberExpression(2)).interpret(CTX) == 5
    def test_zero_raises(self):
        with pytest.raises(ZeroDivisionError): DivideExpression(NumberExpression(5), NumberExpression(0)).interpret(CTX)
    def test_fractional(self): assert DivideExpression(NumberExpression(10), NumberExpression(4)).interpret(CTX) == 2.5

class TestFloorDivideExpression:
    def test_floors(self):     assert FloorDivideExpression(NumberExpression(17), NumberExpression(5)).interpret(CTX) == 3
    def test_zero_raises(self):
        with pytest.raises(ZeroDivisionError): FloorDivideExpression(NumberExpression(5), NumberExpression(0)).interpret(CTX)
    def test_str_contains_slashes(self): assert "//" in str(FloorDivideExpression(NumberExpression(1), NumberExpression(2)))

class TestModuloExpression:
    def test_remainder(self):  assert ModuloExpression(NumberExpression(17), NumberExpression(5)).interpret(CTX) == 2
    def test_zero_raises(self):
        with pytest.raises(ZeroDivisionError): ModuloExpression(NumberExpression(5), NumberExpression(0)).interpret(CTX)
    def test_str_contains_percent(self): assert "%" in str(ModuloExpression(NumberExpression(1), NumberExpression(2)))

class TestPowerExpression:
    def test_power(self):           assert PowerExpression(NumberExpression(2), NumberExpression(10)).interpret(CTX) == 1024
    def test_fractional_exponent(self):
        assert abs(PowerExpression(NumberExpression(4), NumberExpression(0.5)).interpret(CTX) - 2) < 0.001
    def test_str_contains_stars(self): assert "**" in str(PowerExpression(NumberExpression(2), NumberExpression(3)))


# ── Unary functions ───────────────────────────────────────────
class TestNegateExpression:
    def test_negates_positive(self): assert NegateExpression(NumberExpression(5)).interpret(CTX) == -5
    def test_negates_negative(self): assert NegateExpression(NumberExpression(-3)).interpret(CTX) == 3

class TestAbsExpression:
    def test_abs_negative(self): assert AbsExpression(NumberExpression(-9)).interpret(CTX) == 9
    def test_abs_positive(self): assert AbsExpression(NumberExpression(7)).interpret(CTX) == 7

class TestSqrtExpression:
    def test_sqrt(self): assert SqrtExpression(NumberExpression(100)).interpret(CTX) == 10
    def test_negative_raises(self):
        with pytest.raises(ValueError): SqrtExpression(NumberExpression(-1)).interpret(CTX)

class TestInverseExpression:
    def test_inverse(self): assert InverseExpression(NumberExpression(4)).interpret(CTX) == 0.25
    def test_zero_raises(self):
        with pytest.raises(ZeroDivisionError): InverseExpression(NumberExpression(0)).interpret(CTX)

class TestFactorialExpression:
    def test_zero(self):     assert FactorialExpression(NumberExpression(0)).interpret(CTX) == 1
    def test_five(self):     assert FactorialExpression(NumberExpression(5)).interpret(CTX) == 120
    def test_six(self):      assert FactorialExpression(NumberExpression(6)).interpret(CTX) == 720
    def test_negative_raises(self):
        with pytest.raises(ValueError): FactorialExpression(NumberExpression(-1)).interpret(CTX)

class TestPowerOfTenExpression:
    def test_ten_squared(self): assert PowerOfTenExpression(NumberExpression(2)).interpret(CTX) == 100
    def test_ten_to_zero(self): assert PowerOfTenExpression(NumberExpression(0)).interpret(CTX) == 1

class TestLogExpression:
    def test_log_100(self):    assert LogExpression(NumberExpression(100)).interpret(CTX) == 2
    def test_log_1(self):      assert LogExpression(NumberExpression(1)).interpret(CTX) == 0
    def test_zero_raises(self):
        with pytest.raises(ValueError): LogExpression(NumberExpression(0)).interpret(CTX)
    def test_negative_raises(self):
        with pytest.raises(ValueError): LogExpression(NumberExpression(-5)).interpret(CTX)


# ── Composite trees ───────────────────────────────────────────
class TestCompositeTrees:
    def test_diff_of_squares(self):
        a, b = VariableExpression("a"), VariableExpression("b")
        ctx = Context().set("a", 10).set("b", 5)
        expr = MultiplyExpression(AddExpression(a, b), SubtractExpression(a, b))
        assert expr.interpret(ctx) == 75

    def test_sqrt_abs_composition(self):
        a, b = VariableExpression("a"), VariableExpression("b")
        ctx = Context().set("a", 7).set("b", 23)
        expr = SqrtExpression(AbsExpression(SubtractExpression(a, b)))
        assert expr.interpret(ctx) == 4

    def test_same_tree_different_contexts(self):
        a, b = VariableExpression("a"), VariableExpression("b")
        expr = MultiplyExpression(AddExpression(a, b), SubtractExpression(a, b))
        assert expr.interpret(Context().set("a", 10).set("b", 5)) == 75
        assert expr.interpret(Context().set("a", 6).set("b", 2))  == 32

    def test_pow10_abs_composition(self):
        b = VariableExpression("b")
        ctx = Context().set("b", -3)
        assert PowerOfTenExpression(AbsExpression(b)).interpret(ctx) == 1000
